### PR TITLE
Remaining flight range based RTL trigger

### PR DESCRIPTION
**Pull request by jkflying**
**Describe problem solved by this pull request**
Currently the RTL is triggered by a fixed battery level, irrespective of the distance to home, wind conditions, vehicle flight speed etc. This means that the vehicle could run out of battery during a long RTL, or arrive home with still a large fraction of battery remaining during a short RTL.

**Describe your solution**
Take into account the flight time home during the RTL. From the RTL viability checks in navigator publish this information saying how long the RTL will take (taking into account vehicle speeds, wind speed and destination distance/direction), and what fraction of the total battery would be used by the RTL. In Commander, subscribe to this and trigger RTL if the current battery level less the battery used on RTL will take the vehicle below the existing low battery thresholds.

**Describe possible alternatives**
A parameter was used for "total flight time in RTL". This could have been taken from the power module / smart battery telemetry of the current and battery capacity values. However, for multirotor that is risky because power usage in forward flight is higher than hover, so a vehicle hovering far away would RTL too late. The parameter approach also lets users set a hard upper limit on the amount of time they would be happy with the vehicle spending in RTL, which might be useful for regulatory or safety reasons.

**Test data / coverage**
This was tested extensively in Auterion, but I haven't tested it yet since upstreaming. It should get some SITL time to make sure nothing broke in the transplant, but the concept is quite solid.

Note: it might be better to squash this on merge

### CODE REVIEW

File: src/modules/navigator/rtl_params.c [hunk 1]
```diff
@@ -136,6 +136,16 @@ PARAM_DEFINE_INT32(RTL_TYPE, 0);
  */
 PARAM_DEFINE_INT32(RTL_CONE_ANG, 0);
 
+/**
+ * Maximum allowed RTL flight in minutes
+ *
+ * This is used to determine when the vehicle should be switched to RTL due to low battery.
+ * Note, particularly for multirotors this should reflect flight time at cruise speed, not while stationary
+ *
+ * @group Commander
```
Comment (dagar) [inline src/modules/navigator/rtl_params.c:145]: units

File: src/modules/navigator/rtl.h [hunk 2]
```diff
@@ -141,6 +149,20 @@ class RTL : public MissionBlock, public ModuleParams
 		(ParamFloat<px4::params::RTL_MIN_DIST>) _param_rtl_min_dist,
 		(ParamInt<px4::params::RTL_TYPE>) _param_rtl_type,
 		(ParamInt<px4::params::RTL_CONE_ANG>) _param_rtl_cone_half_angle_deg,
+		(ParamFloat<px4::params::RTL_FLT_TIME>) _param_rtl_flt_time,
 		(ParamInt<px4::params::RTL_PLD_MD>) _param_rtl_pld_md
 	)
+
+	// These need to point at different parameters depending on vehicle type.
+	// Can't hard-code them because we have non-MC/FW/Rover builds
+	uint8_t _rtl_vehicle_type{255};
```
Comment (dagar) [inline src/modules/navigator/rtl.h:158]: vehicle_status_s::VEHICLE_TYPE_UNKNOWN = 0

File: src/modules/navigator/rtl.h [hunk 3]
```diff
@@ -141,6 +149,20 @@ class RTL : public MissionBlock, public ModuleParams
...
+	param_t _rtl_xy_speed;
```
Comment (dagar) [inline src/modules/navigator/rtl.h:159]:
```suggestion
	param_t _rtl_xy_speed{PARAM_INVALID};
```

File: src/modules/navigator/rtl.h [hunk 4]
```diff
@@ -141,6 +149,20 @@ class RTL : public MissionBlock, public ModuleParams
...
+	param_t _rtl_descent_speed;
```
Comment (dagar) [inline src/modules/navigator/rtl.h:160]:
```suggestion
	param_t _rtl_descent_speed{PARAM_INVALID};
```

**Comment by moreba1** [comment]
> A parameter was used for "total flight time in RTL". This could have been taken from the power module

i think "average power consumption" (mah/km) + "distance to home" is better than "flight time in RTL".

https://github.com/PX4/PX4-Autopilot/pull/14819

**Comment by jkflying** [comment]
> I think uisng "average power consumption" (mah/km) + "distance to home" is better than "flight time in RTL". #14819

That would be a nice improvement, I agree. However, I'd prefer to get this in with the current feature set, and we can work on improving it once people can already use it.

File: src/modules/navigator/rtl_params.c (hunk 1 above)
Comment (jkflying) [reply src/modules/navigator/rtl_params.c:146]: Is minutes a problem? It is documented in the string. It's just that a number like 15 or 30 is much easier to understand for users than 900 or 1800 if I used seconds instead.

**Comment by jkflying** [comment]
Out of flash again on V2 :cry:

File: src/modules/navigator/rtl_params.c (hunk 1 above)
Comment (dagar) [reply src/modules/navigator/rtl_params.c:146]: Minutes is fine, I mean add the unit metadata (@unit).

**Comment by LorenzMeier** [comment]
Re-running CI

**Review by LorenzMeier — COMMENTED** [review]
Good to go!

**Comment by Antiheavy** [comment]
> > I think uisng "average power consumption" (mah/km) + "distance to home" is better than "flight time in RTL". #14819
> 
> That would be a nice improvement, I agree. However, I'd prefer to get this in with the current feature set, and we can work on improving it once people can already use it.

another potential future future improvement could take into account the wind speed and direction estimate for vehicles that can estimate wind (e.g. those that have airspeed sensors).

However, this PR is a great improvement as a first step as-is.

**Comment by jkflying** [comment]
> take into account the wind speed and direction estimate

It does that already :smile:

**Comment by hamishwillee** [comment]
@jkflying You wouldn't create a PR without making it possible for people to use it - so can you point me to the docs?

Assuming not a few questions:
- Does this work on all vehicle types?
- How do you enable/disable this feature? Is it supported in QGC - how?
- How does this work? My guess is that you set low battery action to RTL. Then you set RTL_FLT_TIME to indicate max time expected for RTL. The low battery RTL will then trigger early based on the systems calculation that time taken to get home at cruise speed will trigger low battery RTL within RTL_FLT_TIME. Is that correct? And presumably you can set this to 0 just to ignore it.

**Pull request merged**